R/kwic.R

```diff
@@ -135,39 +135,84 @@ is.kwic <- function(x) {
     inherits(x, "kwic")
 }
 
+#' @rdname kwic
+#' @method as.data.frame kwic
+#' @export
+as.data.frame.kwic <- function(x, ..., window = NULL, separator = NULL) {
+    
+    if (is.null(window))
+        window <- attr(x, "window")
+    else
+        window <- check_integer(window, 1, 1, 0)
+    
+    if (is.null(separator))
+        separator <- attr(x, "separator")
+    else
+        separator <- check_character(separator)
+    
+    attrs <- attributes(x)
+    x$pre <- rep("", nrow(x))
+    x$keyword <- rep("", nrow(x))
+    x$post <- rep("", nrow(x))
+    if (nrow(x)) {
+        toks <- attrs$tokens[x$docname]
+        lis_pre <- as.list(tokens_select(toks, startpos = pmax(x$from - window, 1), endpos = x$from - 1))
+        lis_key <- as.list(tokens_select(toks, startpos = x$from, endpos = x$to))
+        lis_post <- as.list(tokens_select(toks, startpos = x$to + 1, endpos = x$to + window))
+        
+        x$pre[lengths(lis_pre) > 0] <- stri_c_list(lis_pre, sep = separator)
+        x$keyword[lengths(lis_key) > 0] <- stri_c_list(lis_key, sep = separator)
+        x$post[lengths(lis_post) > 0] <- stri_c_list(lis_post, sep = separator)
+    }
+    attr(x, "tokens") <- NULL
+    attr(x, "window") <- NULL
+    attr(x, "separator") <- NULL
+    class(x) <- "data.frame"
+    # reorder columns to match pre-v3 order
+    x <- x[, c("docname", "from", "to", "pre", "keyword", "post", "pattern")]
+    return(x)
+}
+
+#' @rdname kwic
 #' @method print kwic
-#' @importFrom stringi stri_c
-#' @noRd
+#' @param window the number of context words to be displayed around the keyword.
+#' @param separator character to separate words in the output
+#' @importFrom stringi stri_c stri_c_list
 #' @export
-print.kwic <- function(x, ...) {
-    if (!nrow(x)) {
-        cat("kwic object with 0 rows")
-    } else if (!is.kwic(x)) {
-        NextMethod()
-    } else {
+print.kwic <- function(x, window = NULL, separator = NULL, ...) {
+    
+    attrs <- attributes(x)
+    x <- as.data.frame(x, window = window, separator = separator)
+    
+    cat(sprintf("Keyword-in-context with %d match%s.", nrow(x), ifelse(nrow(x) != 1, "es", "")))
```

Comment:
Don't you want to use `quanteda:::msg()`?

Reply:
Sure, that's fine, but you wrote that line! I almost never use `sprintf()`.